components/feature/downloads/src/main/java/mozilla/components/feature/downloads/share/ShareDownloadFeature.kt

```diff
@@ -0,0 +1,244 @@
+ *
+ * with a 1 second timeout to ensure a smooth UX.
+ *
+ * To finish the process in this small timeframe the feature is recommended to be used only for images.
+ *
+ *  @param context Android context used for various platform interactions
+ *  @param httpClient Client used for downloading internet resources
+ *  @param store a reference to the application's [BrowserStore]
+ */
+@Suppress("TooManyFunctions")
+class ShareDownloadFeature(
+    private val context: Context,
+    private val httpClient: Client,
+    private val store: BrowserStore,
+    private val tabId: String?
+) : LifecycleAwareFeature {
+    private val logger = Logger("ShareDownloadMiddleware")
+
+    @VisibleForTesting
+    internal var scope: CoroutineScope? = null
+
+    override fun start() {
+        scope = store.flowScoped { flow ->
+            flow.mapNotNull { state -> state.findTabOrCustomTabOrSelectedTab(tabId) }
+                .ifChanged { it.content.share }
+                .collect { state ->
+                    state.content.share?.let { shareState ->
+                        logger.debug("Starting the sharing process")
+                        startSharing(shareState)
+
+                        // This is a fire and forget action, not something that we want lingering the tab state.
+                        store.dispatch(ShareInternetResourceAction.ConsumeShareAction(state.id))
+                    }
+                }
+        }
+    }
+
+    override fun stop() {
+        scope?.cancel()
+    }
+
+    init {
+        CoroutineScope(cleanupCacheCoroutineDispatcher).launch {
+            cleanupCache()
+        }
+    }
+
+    @VisibleForTesting
+    internal fun startSharing(internetResource: ShareInternetResourceState) {
+        val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable ->
+            when (throwable) {
+                is InterruptedException -> {
+                    logger.warn("Share failed: operation timeout reached")
+                }
+                is IOException,
+                is RuntimeException,
+                is NullPointerException -> {
+                    logger.warn("Share failed: $throwable")
+                }
+            }
+        }
+
+        scope?.launch(coroutineExceptionHandler) {
+            withTimeout(OPERATION_TIMEOUT_MILLIS) {
+                val download = download(internetResource)
+                share(
+                    contentType = internetResource.contentType,
+                    filePath = download.canonicalPath,
+                    message = sanitizeLongUrls(internetResource.url)
+                )
+            }
+        }
+    }
+
+    @WorkerThread
+    @VisibleForTesting
+    internal fun download(internetResource: ShareInternetResourceState): File {
+        val cookiePolicy = if (internetResource.private) {
+            Request.CookiePolicy.OMIT
+        } else {
+            Request.CookiePolicy.INCLUDE
+        }
```

Comment:
Sorry just to let you know that we updated our version of A-S and we brought back the api https://github.com/mozilla-mobile/android-components/pull/9557

Reply:
Thank you.
Moved to use the `private` property instead of `cookiesPolicy`